ofrak_core/ofrak/core/free_space.py

```diff
@@ -387,176 +382,97 @@ class FreeSpaceModifier(Modifier[FreeSpaceModifierConfig]):
-
-        await mem_region_view.resource.create_child_from_view(
-            FreeSpace(
-                virtual_patch_range.start,
-                virtual_patch_range.length(),
-                config.permissions,
-            ),
-            data_range=patch_range,
-            data=nop_bytes,
+        start = mem_region_view.virtual_address
+        size = mem_region_view.size
+        if config.subrange:
+            if len(config.stub) > config.subrange.length():
+                raise ValueError(
+                    f"The provided stub ({len(config.stub)} bytes) cannot be larger than the "
+                    f"Resource subrange @ {hex(mem_region_view.virtual_address)} being freed "
+                    f"({config.subrange.length()} bytes)."
+                )
+            else:
+                start = config.subrange.start
+                size = config.subrange.length()
+
+        parent_mr_view = await resource.get_parent_as_view(MemoryRegion)
+        stub_range = Range.from_size(start, len(config.stub))
+        stub_offset = parent_mr_view.get_offset_in_self(stub_range.start)
+        stub_offset_range = stub_range.translate(stub_offset - stub_range.start)
+
+        freed_range = Range(start + stub_range.length(), start + size)
+        freed_offset = parent_mr_view.get_offset_in_self(freed_range.start)
+        freed_offset_range = freed_range.translate(freed_offset - freed_range.start)
+
+        if config.subrange:
+            await self._find_and_delete_overlapping_children(resource, freed_range)
+
+        # If a stub is provided we only assume this is a function if the virtual
+        # address of `resource` is the same value as config.subrange.start.
+        name = None
+        is_bb = False
+        if resource.has_tag(ComplexBlock) and start == mem_region_view.virtual_address:
+            cb_view = await resource.view_as(ComplexBlock)
+            name = str(cb_view.Symbol)
+        elif start > mem_region_view.virtual_address:
+            # our subrange begins within an existing resource... so let's try making it a basic block
+            is_bb = True
+
+        # We need to delete the containing resource before we can begin creating new Memory Region
+        # resources within
+        await resource.delete()
+        await resource.save()
+
+        # indicating we know this is a CB and we want to preserve our knowledge of linkage
+        if stub_range.length() > 0:
+            if name:
+                await parent_mr_view.resource.create_child_from_view(
+                    ComplexBlock(stub_range.start, stub_range.end, name),
+                    data_range=stub_offset_range,
+                    data=config.stub,
+                )
+            elif is_bb:
+                await parent_mr_view.resource.create_child_from_view(
+                    BasicBlock(
+                        stub_range.start, stub_range.length(), InstructionSetMode.NONE, True, None
+                    ),
```

Comment:
I question whether or not `BasicBlock` should need to know the `InstructionSetMode`, but imo it's out of scope of this set of changes.